index.js

```diff
@@ -0,0 +1,282 @@
+                    this.createSharedChunk(
+                        compilation,
+                        allChunks,
+                        globalModulesSet,
+                        sharedChunksSet,
+                        spec
+                    );
+                }
+            });
+
+            // additional-chunk-assets is the only unconditional hook that gets
+            // passed all of the chunks that is called after sortItemsWithModuleIds
+            // and sortItemsWithChunkIds which are two internal webpack methods
+            // in Compilation.js that sort the order of the parents based on
+            // chunk id.
+            compilation.plugin(['additional-chunk-assets'], allChunks => {
+                for (const chunk of allChunks) {
+                    // Sort dependencies based on the order in which they should
+                    // be loaded.
+                    chunk.parents = chunk.parents.sort(
+                        (a, b) => a.parents.includes(b) ? 1 : -1);
+                }
+            });
+        });
+    }
+
+    // Create a shared chunk
+    //
+    // The process involes the following steps:
+    // - determine which modules should be added to the new shared chunk
+    //   - if any of these module appear in an existing shared chunk, add that
+    //     chunk as a dependency of the new chunk
+    // - remove those modules from other chunks
+    // - add those modules to the new chunk
+    createSharedChunk(
+        compilation,
+        allChunks,
+        globalModulesSet,
+        sharedChunksSet,
+        spec
+    ) {
+        // Create a map between chunks and their names.  The name corresponds
+        // to either the name in the spec (shared chunks) or the name in the
+        // 'entry' part of a webpack.config.js file.
+        const allChunksNameMap = allChunks.reduce((map, chunk) => {
+            if (chunk.name) {
+                map.set(chunk.name, chunk);
+            }
+            return map;
+        }, new Map());
+
+        // Affected chunks will have one or more modules removed from them.
+        // This will also include new chunks generated from previous calls to
+        // processSharedSpec.
+        const affectedChunks = spec.selectedChunks
+            ? allChunks.filter(chunk => {
+                  return spec.selectedChunks.includes(chunk.name);
+              })
+            : allChunks;
+
+        // Track how chunks each module appears in.
+        const commonModulesToCountMap = new Map();
+        for (const chunk of affectedChunks) {
+            for (const module of chunk.modulesIterable) {
+                // TODO(kevinb): add an option to also exclude modules that are
+                // already part of globalModulesSet, currently our filters
+                // produce a disjoint set of modules... or at the very least
+                // warn when there an overlap in the set of entry modules
+                // between shared chunks.
```

Comment:
I'm going to add the warning before merging and will open an issue to track the TODO.